icu4j/main/classes/core/src/com/ibm/icu/impl/personname/PersonNameFormatterImpl.java

```diff
@@ -276,65 +307,73 @@ private PersonNamePattern getBestPattern(PersonNamePattern[] patterns, PersonNam
-            // locale, we use the formatter's locale as the name's locale
-            int formatterScript = UScript.getCodeFromName(ULocale.addLikelySubtags(ULocale.forLocale(locale)).getScript());
-            String givenName = name.getFieldValue(PersonName.NameField.GIVEN, new HashSet<PersonName.FieldModifier>());
-            int nameScript = UScript.INVALID_CODE;
-            for (int i = 0; nameScript == UScript.INVALID_CODE && i < givenName.length(); i++) {
-                // the script of the name is the script of the first character in the name whose script isn't
-                // COMMON or INHERITED
-                int script = UScript.getScript(givenName.charAt(i));
-                if (script != UScript.COMMON && script != UScript.INHERITED) {
-                    nameScript = script;
-                }
-            }
-            if (formatterScript == nameScript) {
-                nameLocale = this.locale;
-            } else {
-                // if the name's script is different from the formatter's script, we use addLikelySubtags() to find the
-                // default language for the name's script and use THAT as the name's locale
-                nameLocale = new Locale(ULocale.addLikelySubtags(new ULocale("und_" + UScript.getShortName(nameScript))).getLanguage());
-            }
-            // TODO: This algorithm has a few deficiencies: First, it assumes the script of the string is the script of the first
-            // character in the string that's not COMMON or INHERITED.  This won't work well for some languages, such as Japanese,
-            // that use multiple scripts.  Doing better would require adding a new getScript(String) method on UScript, which
-            // might be something we want.  Second, we only look at the given-name field.  This field should always be populated,
-            // but if it isn't, we're stuck.  Looking at all the fields requires API on PersonName that we don't need anywhere
-            // else.
+        String nameScript = getNameScript(name);
+        String nameLanguage;
+        Locale specifiedNameLocale = name.getNameLocale();
+
+        Locale.Builder builder = new Locale.Builder();
+        if (specifiedNameLocale != null) {
+            builder.setLocale(specifiedNameLocale);
+            nameLanguage = specifiedNameLocale.getLanguage();
+        } else {
+            ULocale tmpLocale = ULocale.addLikelySubtags(new ULocale("und_" + nameScript));
+            nameLanguage = tmpLocale.getLanguage();
+            builder.setLanguage(nameLanguage);
         }
-        return nameLocale;
+
+        // HACK to work around the fact that "Hani" isn't used in the IDs of any resource bundles-- map it to the
+        // default scripts for the relevant language codes
+        Map<String, String> SCRIPT_OVERRIDES = new HashMap<String, String>() {{
+            put("zh-Hani", "Hans");
+            put("ja-Hani", "Jpan");
+            put("ko-Hani", "Kore");
```

Comment:
Should also include ko-Hang, ja-Kana, ja-Hira. They will probably not come up, but just in case. Also, I haven't checked, but are you dealing with old-style ICU locales (like zh_Hani) or BCP47 style (like zh-Hani).

I think it might be better to match in the other direction for Hani, because the resource bundles will contain both zh_Hans and zh_Hant

Reply:
>Also, I haven't checked, but are you dealing with old-style ICU locales (like zh_Hani) or BCP47 style (like zh-Hani).

In this table, it's neither, really.  I need to look up based on both the language and the script, so I combine them into one key by concatenating them and putting a hyphen in between.  It looks like a BCP47 language tag, but it's really just an arbitrary string to facilitate lookup in a hash map.

>I think it might be better to match in the other direction for Hani, because the resource bundles will contain both zh_Hans and zh_Hant

I don't know what you mean by "match in the other direction."  If I get "Hani" as my script, it means I can't distinguish between "Hans" and "Hant", so I arbitrarily picked "Hant".  My thinking was that for the purposes of finding out if locale data for the locale ID exists, it doesn't matter, because we know both zh-Hans and zh-Hant exist.